packages/state/src/lib/transactions.ts

```diff
@@ -300,3 +310,61 @@ export function transact<T>(fn: () => T): T {
 	}
 	return transaction(fn)
 }
+
+/**
+ * @internal
+ */
+export async function deferAsyncEffects<T>(fn: () => Promise<T>) {
+	// Can't kick off async transactions during a sync transaction because
+	// the async transaction won't finish until after the sync transaction
+	// is done.
+	if (inst.currentTransaction?.isSync) {
+		throw new Error('deferAsyncEffects cannot be called during a sync transaction')
+	}
+
+	// Can't kick off async transactions during a reaction phase at the moment,
+	// because the transaction stack is cleared after the reaction phase.
+	// So wait until the path ahead is clear
+	while (inst.globalIsReacting) {
+		await sleep(0)
```

Comment:
tiniest of nits but maybe a microtask would be more appropriate than a timeout here? i could also see this just throwing and forcing the callee to handle this, the initial call of the callback being async in some conditions could be surprising